README.md
# Flow XO Stripe Example Service

This is a Stripe Example service module for the [Flow XO](https://flowxo.com) platform. For more details on how to develop and test this service, please refer to the [Flow XO SDK](http://github.com/flowxo/flowxo-sdk).

## Usage

Steps to run the service from the command line using the Flow XO SDK
``` bash
# Clone the repo
git clone https://github.com/flowxo/flowxo-services-stripe-example
cd flowxo-services-stripe-example

# Install the dependencies
npm install
grunt init


# Create a .env file with the following content (no hashes):
STRIPE_EXAMPLE_ID=<YOUR_APP_ID>
STRIPE_EXAMPLE_SECRET=<YOUR_APP_SECRET>

# Create an authentication
grunt auth

# Run methods
grunt run
```

## Contributing

``` bash
# Clone the repo
git clone https://github.com/flowxo/flowxo-services-stripe-example

# Install the dependencies
npm install -g yo grunt-cli
npm install
grunt init

# Generate a new method
yo flowxo:method

# Watch files for changes and running style checks on change
grunt

# Create/renew an authentication file
grunt auth

# Run integration tests, using authentication
grunt run [--record --replay --name=<name>]

```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0 - 2015-08-19
### Added
- Initial release.


lib/index.js
'use strict';

var sdk = require('flowxo-sdk');

var service = new sdk.Service({
  serviceRoot: __dirname,
  name: 'Stripe',
  slug: 'stripe',
  auth: {
    type: 'oauth2',

    // The strategy key should return the Passport Strategy for your service.
    // Passport has numerous available strategies for popular services, normally named
    // e.g. passport-facebook, passport-twitter etc
    // Example:
    // strategy: require('passport-facebook').Strategy
    strategy: require('passport-stripe').Strategy,

    // These options will be passed to the strategy when registering.
    // An OAuth 2.0 strategy requires `clientID` and `clientSecret`
    // to be passed. Fill in your ID and secret for this service in the
    // .env file and they will be populated at runtime below.
    // If your strategy requires any other options to be passed when registering,
    // add them below.
    options: {
      clientID: process.env.STRIPE_ID,
      clientSecret: process.env.STRIPE_SECRET,
      state: true
    },

    // Authentication parameters to be used.
    // These are sent when making an OAuth request.
    // For example, where an OAuth 2.0 API defines access scopes,
    // you may send
    // params:{
    //   scope: ['allow_email']
    // }
    params: {
      scope: ['read_write']
    }
  },
  scripts: {

  }
});

service.Client = require('./client');

module.exports = service;


lib/client.js
'use strict';

var stripe = require('stripe')();

var StripeClient = function(credentials, service) {
  this.service = service;
  this.credentials = credentials;
  this.stripe = stripe;
  this.stripe.setTimeout(30000);
};

var isoifyDateField = function(field) {
  return new Date(field * 1000).toISOString();
};

var convertDateFields = function(field, props) {
  props.forEach(function(prop) {
    if(field[prop]) {
      field[prop] = isoifyDateField(field[prop]);
    }
  });
  return field;
};

StripeClient.prototype.getCustomers = function() {
  return stripe.customers.list({
    limit: 10
  }, this.credentials.access_token).then(function(response) {
    return response.data.map(function(field) {
      convertDateFields(field, ['created']);
      return field;
    });
  });

};



StripeClient.prototype.getCustomer = function(id) {
  return stripe.customers
    .retrieve(id, this.credentials.access_token)
    .then(function(customer) {
      if(customer.created) {
        customer.created = isoifyDateField(customer.created);
      }
      return customer;
    });
};

module.exports = StripeClient;


lib/methods/new_customer/config.js
'use strict';

var config = {
  name: 'New Customer',
  slug: 'new_customer',
  type: 'poller',
  kind: 'trigger',
  scripts: {
    run: require('./run')
  },
  fields: {
    input: [],
    output: [{
      key: 'object',
      label: 'Object'
    }, {
      key: 'created',
      label: 'Created At'
    }, {
      key: 'id',
      label: 'ID'
    }, {
      key: 'livemode',
      label: 'LiveMode'
    }, {
      key: 'description',
      label: 'Description'
    }, {
      key: 'email',
      label: 'Email'
    }, {
      key: 'delinquent',
      label: 'Customer Delinquent'
    }, {
      key: 'subscriptions__total_count',
      label: 'Subscription Count'
    }, {
      key: 'subscriptions__url',
      label: 'Subscriptions Url'
    }, {
      key: 'subscriptions__data__0__id',
      label: 'Subscription ID'
    }, {
      key: 'subscriptions__data__0__plan__interval',
      label: 'Subscription Plan Interval'
    }, {
      key: 'subscriptions__data__0__plan__name',
      label: 'Subscription Plan Name'
    }, {
      key: 'subscriptions__data__0__plan__created',
      label: 'Subscription Plan Created On'
    }, {
      key: 'subscriptions__data__0__plan__amount',
      label: 'Subscription Plan Amount'
    }, {
      key: 'subscriptions__data__0__plan__currency',
      label: 'Subscription Plan Currency'
    }, {
      key: 'subscriptions__data__0__plan__id',
      label: 'Subscription Plan ID'
    }, {
      key: 'subscriptions__data__0__plan__object',
      label: 'Subscription Plan Object'
    }, {
      key: 'subscriptions__data__0__plan__livemode',
      label: 'Subscription Plan LiveMode'
    }, {
      key: 'subscriptions__data__0__plan__interval_count',
      label: 'Subscription Plan Interval Count'
    }, {
      key: 'subscriptions__data__0__plan__trial_period_days',
      label: 'Subscription Plan Trial Period Days'
    }, {
      key: 'subscriptions__data__0__plan__statement_descriptor',
      label: 'Subscription Plan Statement Description'
    }, {
      key: 'subscriptions__data__0__object',
      label: 'Subscription Plan Object'
    }, {
      key: 'subscriptions__data__0__start',
      label: 'Subscription Plan Start'
    }, {
      key: 'subscriptions__data__0__status',
      label: 'Subscription Plan Status'
    }, {
      key: 'subscriptions__data__0__customer',
      label: 'Subscription Plan Customer'
    }, {
      key: 'subscriptions__data__0__cancel_at_period_end',
      label: 'Subscription Plan Cancel At Period End'
    }, {
      key: 'subscriptions__data__0__current_period_start',
      label: 'Subscription Plan Period Start'
    }, {
      key: 'subscriptions__data__0__current_period_end',
      label: 'Subscription Plan Period End'
    }, {
      key: 'subscriptions__data__0__ended_at',
      label: 'Subscription Plan Ended At'
    }, {
      key: 'subscriptions__data__0__trial_start',
      label: 'Subscription Plan Trial Start'
    }, {
      key: 'subscriptions__data__0__trial_end',
      label: 'Subscription Plan Trial End'
    }, {
      key: 'subscriptions__data__0__canceled_at',
      label: 'Subscription Plan Cancelled At'
    }, {
      key: 'subscriptions__data__0__quantity',
      label: 'Subscription Plan Quantity'
    }, {
      key: 'subscriptions__data__0__application_fee_percent',
      label: 'Subscription Plan Application Fee Percent'
    }, {
      key: 'subscriptions__data__0__discount',
      label: 'Subscription Plan Discount'
    }, {
      key: 'subscriptions__data__0__tax_percent',
      label: 'Subscription Plan Tax Percent'
    }, {
      key: 'discount',
      label: 'Discount'
    }, {
      key: 'account_balance',
      label: 'Account Balance'
    }, {
      key: 'currency',
      label: 'Currency'
    }, {
      key: 'sources__total_count',
      label: 'Total Sources'
    }, {
      key: 'sources__url',
      label: 'Payment Sources Url'
    }, {
      key: 'sources__data__0__id',
      label: 'Active Card ID'
    }, {
      key: 'sources__data__0__object',
      label: 'Active Card Object'
    }, {
      key: 'sources__data__0__last4',
      label: 'Active Card Last4'
    }, {
      key: 'sources__data__0__brand',
      label: 'Active Card Type'
    }, {
      key: 'sources__data__0__funding',
      label: 'Active Card Funding'
    }, {
      key: 'sources__data__0__exp_month',
      label: 'Active Card Exp Month'
    }, {
      key: 'sources__data__0__exp_year',
      label: 'Active Card Exp Year'
    }, {
      key: 'sources__data__0__fingerprint',
      label: 'Active Card Fingerprint'
    }, {
      key: 'sources__data__0__country',
      label: 'Active Card Country'
    }, {
      key: 'sources__data__0__name',
      label: 'Active Card Name'
    }, {
      key: 'sources__data__0__address_line1',
      label: 'Active Card Address Line1'
    }, {
      key: 'sources__data__0__address_line2',
      label: 'Active Card Address Line2'
    }, {
      key: 'sources__data__0__address_city',
      label: 'Active Card City'
    }, {
      key: 'sources__data__0__address_state',
      label: 'Active Card State'
    }, {
      key: 'sources__data__0__address_zip',
      label: 'Active Card Zip'
    }, {
      key: 'sources__data__0__address_country',
      label: 'Active Card Country'
    }, {
      key: 'sources__data__0__cvc_check',
      label: 'Active Card Cvc Checked'
    }, {
      key: 'sources__data__0__address_line1_check',
      label: 'Active Card Address Line1 Checked'
    }, {
      key: 'sources__data__0__address_zip_check',
      label: 'Active Card Zip Checked'
    }, {
      key: 'sources__data__0__customer',
      label: 'Active Card Customer'
    }, {
      key: 'default_source',
      label: 'Default Payment Source'
    }]
  }
};

module.exports = function(service) {
  service.registerMethod(config);
};


Gruntfile.js
// Generated on 2015-08-19 using generator-flowxo 1.0.5
'use strict';

module.exports = function(grunt) {

  // Fix grunt options
  // Can remove if using grunt 0.5
  // https://github.com/gruntjs/grunt/issues/908
  require('nopt-grunt-fix')(grunt);

  // Load grunt tasks automatically
  require('load-grunt-tasks')(grunt);

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);

  // Load the Flow XO tasks
  grunt.loadNpmTasks('flowxo-sdk');

  // Define the configuration for all the tasks
  grunt.initConfig({
    env: {
      src: '.env'
    },
    watch: {
      js: {
        options: {
          spawn: true,
          interrupt: false,
          debounceDelay: 250
        },
        files: ['lib/**/*.js'],
        tasks: ['jshint']
      }
    },
    jshint: {
      options: {
        jshintrc: true,
        reporter: require('jshint-stylish')
      },
      source: {
        src: ['Gruntfile.js', 'lib/**/*.js']
      }
    },
    jsbeautifier: {
      options: {
        config: '.jsbeautifyrc'
      },
      source: {
        src: ['Gruntfile.js', 'lib/**/*.js']
      }
    },
    flowxo: {
      options: {
        credentialsFile: 'credentials.json',
        getService: function() {
          return require('./lib');
        },
      },
      init: {},
      auth: {},
      run: {
        options: {
          runsFolder: 'runs',
          webhookPort: 9095
        }
      }
    }
  });

  // Initialise the SDK
  grunt.registerTask('init', ['flowxo:init']);

  // Authentication Tasks
  grunt.registerTask('auth', ['env', 'flowxo:auth']);

  // Run Tasks
  grunt.registerTask('run', ['env', 'flowxo:run']);

  // Preflight Tasks
  grunt.registerTask('preflight', ['env', 'jsbeautifier', 'jshint']);

  // Default Task
  grunt.registerTask('default', ['watch']);
};
